File: C/TCS34725.c
// Distributed with a free-will license.
// Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
// TCS34725
// This code is designed to work with the TCS34725_I2CS I2C Mini Module available from ControlEverything.com.
// https://www.controleverything.com/products

#include <stdio.h>
#include <stdlib.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <fcntl.h>

void main() 
{
	// Create I2C bus
	int file;
	char *bus = "/dev/i2c-1";
	if ((file = open(bus, O_RDWR)) < 0) 
	{
		printf("Failed to open the bus. \n");
		exit(1);
	}
	// Get I2C device, TCS34725 I2C address is 0x29(41)
	ioctl(file, I2C_SLAVE, 0x29);

	// Select enable register(0x80)
	// Power ON, RGBC enable, wait time disable(0x03)
	char config[2] = {0};
	config[0] = 0x80;
	config[1] = 0x03;
	write(file, config, 2);
	// Select ALS time register(0x81)
	// Atime = 700 ms(0x00)
	config[0] = 0x81;
	config[1] = 0x00;
	write(file, config, 2);
	// Select Wait Time register(0x83)
	// WTIME : 2.4ms(0xFF)
	config[0] = 0x83;
	config[1] = 0xFF;
	write(file, config, 2);
	// Select control register(0x8F)
	// AGAIN = 1x(0x00)
	config[0] = 0x8F;
	config[1] = 0x00;
	write(file, config, 2);
	sleep(1);

	// Read 8 bytes of data from register(0x94)
	// cData lsb, cData msb, red lsb, red msb, green lsb, green msb, blue lsb, blue msb
	char reg[1] = {0x94};
	write(file, reg, 1);
	char data[8] = {0};
	if(read(file, data, 8) != 8)
	{
		printf("Erorr : Input/output Erorr \n");
	}
	else
	{
		// Convert the data
		int cData = (data[1] * 256 + data[0]);
		int red = (data[3] * 256 + data[2]);
		int green = (data[5] * 256 + data[4]);
		int blue = (data[7] * 256 + data[6]);

		// Calculate luminance
		float luminance = (-0.32466) * (red) + (1.57837) * (green) + (-0.73191) * (blue);
		if(luminance < 0)
		{
			luminance = 0;
		}

		// Output data to screen
		printf("Red color luminance : %d lux \n", red);
		printf("Green color luminance : %d lux \n", green);
		printf("Blue color luminance : %d lux \n", blue);
		printf("IR  luminance : %d lux \n", cData);
		printf("Ambient Light Luminance : %.2f lux \n", luminance);
	}
}

File: Onion Omega Python/TCS34725.py
# Distributed with a free-will license.
# Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
# TCS34725
# This code is designed to work with the TCS34725_I2CS I2C Mini Module available from ControlEverything.com.
# https://www.controleverything.com/products

from OmegaExpansion import onionI2C
import time

# Get I2C bus
i2c = onionI2C.OnionI2C()

# TCS34725 address, 0x29(41)
# Select Enable register, 0x80(128)
#		0x03(03)	Power ON, RGBC enable, RGBC Interrupt Mask not asserted
#					Wait disable, Sleep After Interrupt not asserted
i2c.writeByte(0x29, 0x80, 0x03)
# TCS34725 address, 0x29(41)
# Select RGBC Timing register, 0x81(129)
#		0x00(00)	ATIME : 700ms
i2c.writeByte(0x29, 0x81, 0x00)
# TCS34725 address, 0x29(41)
# Select Wait Time register, 0x83(131)
#		0xFF(255)	WTIME : 2.4ms
i2c.writeByte(0x29, 0x83, 0xFF)
# TCS34725 address 0x29(41)
# Select Control register, 0x8F(143)
#		0x00(00)	AGAIN is 1x
i2c.writeByte(0x29, 0x8F, 0x00)

# TCS34725 address 0x29(41)
# Read data back from 0x94(148), 8 bytes
# cData LSB, cData MSB, Red LSB, Red MSB, Green LSB, Green MSB
# Blue LSB, Blue MSB 
data = i2c.readBytes(0x29, 0x94, 8)

# Convert the data
cData = data[1] * 256 + data[0]
red = data[3] * 256 + data[2]
green = data[5] * 256 + data[4]
blue = data[7] * 256 + data[6]

# Calculate luminance
luminance = (-0.32466 * red) + (1.57837 * green) + (-0.73191 * blue)

# Output data to screen
print "Red Color Luminance : %d lux" %red
print "Green Color Luminance : %d lux" %green
print "Blue Color Luminance : %d lux" %blue
print "InfraRed Luminance : %d lux" %cData
print "Ambient Light Luminance : %.2f lux" %luminance
